### ISSUE TITLE
Add declare-good

### ISSUE THREAD

**Issue by maksiks** [rucio/rucio#6001]
### Description

Hi Everyone, I encountered a problem that happens at 1% level when we try to transfer data to T2_US_MIT_Tape. Approximately 1% of the files are declared as stuck, however when I compare filesize and checksum between source and destination they match. Hence to me it looks like the transfer went fine, but for whatever reason  this was not communicated back correctly to the FTS. Hence I would like to have an option that would allow me as a site admin to communicate back to Rucio: "I checked that file, all looks good, please mark it as good'. 
It looks like  already exist, please consider adding declare-good option?

Thank you, --Max

### Motivation

Will allow to fix stuck files that were transferred fine but declared as stuck. 

### Change

add declare-good option to rucio-admin replicas declare-bad

**Comment by mlassnig** [comment]
pinging @mpatrascoiu because this an effect of FTS/TAPE interaction

**Comment by mpatrascoiu** [comment]
Hello,

We're investigating the same issue (but over SNow discussion).

Summary of the problem:
Lets's take this file as our example: `/store/data/Run2022D/ParkingDoubleMuonLowMass0/RAW/v1/000/357/899/00000/f5f1d3cb-4fa6-4778-a22a-895c10af824a.root` ([Monit-Kibana](https://monit-kibana.cern.ch/kibana/goto/84c99cc3e821f698c407965f2922cb18))

The file managed the trasfer part successfully, then fails during Archive Monitoring ([FTS3-CMS link](https://fts3-cms.cern.ch:8449/fts3/ftsmon/#/job/aad586c4-6927-11ed-a81c-fa163ebc953f?file=3501170055), will disappear in ~5 days)
```
job_id = aad586c4-6927-11ed-a81c-fa163ebc953f
file_id = 3501170055
transfer_start_time = 2022-11-23T00:36:06Z
transfer_finish_time = 2022-11-23T00:37:45Z
archiving_start_time = 2022-11-23T00:38:34Z
archiving_finish_time = 2022-11-23T00:38:44Z
error_reason = ARCHIVING [42] Malformed served response.
```

So something was bad in the server response, which made FTS fail the archive monitoring part.

Here is where Rucio & FTS get stuck in a corner-case retry-loop:
\- Rucio reusbmits the file with `dst_file_report: 1`
\- FTS provides the destination file report ([FTS3-CMS link](https://fts3-cms.cern.ch:8449/fts3/ftsmon/#/job/8af147de-6ac7-11ed-b5c2-fa163e1da03c), again disappearing in ~5 days)
```
job_id = 8af147de-6ac7-11ed-b5c2-fa163e1da03c
file_id = 3502606813
"dst_file" : {
    "file_size" : 2153843166,
    "checksum_type" : "ADLER32",
    "checksum_value" : "7a6c902a",
    "file_on_tape" : false,
    "file_on_disk" : true
}
```
\- File integrity passed but has `on_tape: false`, so Rucio will retry this transfer
\- In the end, this creates a wave of retries until finally the file makes it `on_tape: true`

Together with @rcarpa and Katy Ellis, we've discussed this situation and a way to improve it:
1. If the file integrity passes, but `on_tape: false`, Rucio should submit an Archive Monitoring only transfer (`source == destination && archive-timeout > 0`)
2. FTS must allow clients to submit Archive Monitoring only transfers

Cheers,
Mihai

**Comment by rcarpa** [comment]
Can please someone from CMS / FTS verify that T2_US_MIT_Tape does actually supports reporting the needed information when `dst_file_report` is set ? The whole point of this issue and the desire to create a `declare-good` CLI command was, as I have understood (maybe wrongly?) that T2_US_MIT_Tape in CMS has some configuration which results in rucio not asking for `dst_file_report`. Again (what I understood from the previous discussions) is that the site doesn't support the `archive_timeout` option. 

I already asked for this check in the initial discussion on slack and my request was ignored.

**Comment by maksiks** [comment]
I do not know how to answer those questions about T2_US_MIT_Tape as I am in know way member of Rucio team. 
I contacted FTS team to see if they can help, and indeed they found out one of those transfers, and ... the transfer went fine, but at the end FTS expects json that you can see below, however it is not clear what happened but FTS did not receive it (remember that only 99% of transfers go through just fine), hence we are in this situation. 
They claim that the fix is in the works and will happen, but I have to fix this dataset and datasets that we will transfer in the nearest future. One way is to simply delete stuck files and let Rucio retransfer them (I do not like it as I am fully aware what it means for any tape system), or have a CLI that would allow me to do it gracefully.

--Max

xrdfs dtn20.nese.rc.fas.harvard.edu:1094 query prepare 1111 /store/data/Run2022D/ParkingDoubleMuonLowMass0/RAW/v1/000/357/899/00000/f5f1d3cb-4fa6-4778-a22a-895c10af824a.root
{"request_id":"1111","responses":[{"path":"/store/data/Run2022D/ParkingDoubleMuonLowMass0/RAW/v1/000/357/899/00000/f5f1d3cb-4fa6-4778-a22a-895c10af824a.root","path_exists":true,"on_tape":true,"online":true,"requested":true,"has_reqid":true,"req_time":"","error_text":""}]}

**Comment by maksiks** [comment]
also please notice that the response clearly indicates that this file is on tape
--Max

**Comment by mpatrascoiu** [comment]
For @rcarpa , the T2_US_MIT_Tape does support the FTS `--archive-timeout` and `--dst-file-report` feature. Even exemplified by Max above, it accepts `xrdfs query prepare` requests and replies correctly (>99% of the time).

For @maksiks , these files are now stuck in the retry-loop corner-case. They will get unstuck automatically, so you don't need a dedicated `declare-good` CLI tool only for this scenario (of course, there could be other use-cases). I let Rucio experts comment on that further.

Rucio will keep retrying the stuck files (there's a diminishing effect, so time between retries increases more and more) with `--dst-file-report` enabled in FTS. Once the file is on tape (which has already happened), on the next retry, FTS will produce the destination file report, showing it's on tape. 

Example of how it will look like for `root://dtn20.nese.rc.fas.harvard.edu:1094//store/data/Run2022D/ParkingDoubleMuonLowMass0/RAW/v1/000/357/899/00000/f5f1d3cb-4fa6-4778-a22a-895c10af824a.root`:
```json
"file_metadata": {
  "dst_file" : {
    "file_size" : 2153843166,
    "checksum_type" : "ADLER32",
    "checksum_value" : "7a6c902a",
    "file_on_tape" : true,
    "file_on_disk" : true
  }
}
```

Once the file is on tape, with enough time, Rucio will mark it as correct on the next FTS retry. What we want to avoid in the future is the repeated wave of errors until the file makes it on tape.

Cheers,
Mihai

**Comment by maksiks** [comment]
Thank you Mihai, may be I am just impatient. For how long should one wait for Rucio to declare that all is fine? If you check any of these files one would see that all of them respond that they are on tape. So if in a week from now files are in stuck state, would it indicate a problem? 
Thank you, --Max

**Comment by maksiks** [comment]
I agree, lets close this ticket and use another one to keep track of things.
--Max

**Comment by maksiks** [comment]
Oops, wrong ticket, we keep this one open.
--Max

**Comment by maksiks** [comment]
Let me just put here the state of the rule, we will come back to it on Monday to see what heppens.
rucio rule-info 9144dfd0c0f44753a5e8c5ff4d6a9ebd
Id:                         9144dfd0c0f44753a5e8c5ff4d6a9ebd
Account:                    maxi
Scope:                      cms
Name:                       /ParkingDoubleMuonLowMass0/Run2022D-v1/RAW
RSE Expression:             T2_US_MIT_Tape
Copies:                     1
State:                      STUCK
Locks OK/REPLICATING/STUCK: 7775/476/242
Grouping:                   DATASET
Expires at:                 None
Locked:                     True
Weight:                     None
Created at:                 2022-11-17 11:39:30
Updated at:                 2022-11-25 16:47:55
Error:                      RequestErrMsg.TRANSFER_FAILED:DESTINATION [17] Destination file exists and overwrite is not enabled
Subscription Id:            None
Source replica expression:  None
Activity:                   User Subscriptions
Comment:                    None
Ignore Quota:               False
Ignore Availability:        False
Purge replicas:             False
Notification:               NO
End of life:                None
Child Rule Id:              None

**Comment by rcarpa** [comment]
@maksiks, what is the conclusion after the week-end? Did some of these replicas transition to the `Done` state? If yes, we know that https://github.com/rucio/rucio/issues/5975 will fix the issue for good. On the other hand, if none transitioned to Done and all of them are still failing with `Destination file exists and overwrite is not enabled`, more investigation is needed.

**Comment by maksiks** [comment]
Here is how this rule looks like on Monday, as one can see no files moved into done state.
rucio rule-info 9144dfd0c0f44753a5e8c5ff4d6a9ebd
Id:                         9144dfd0c0f44753a5e8c5ff4d6a9ebd
Account:                    maxi
Scope:                      cms
Name:                       /ParkingDoubleMuonLowMass0/Run2022D-v1/RAW
RSE Expression:             T2_US_MIT_Tape
Copies:                     1
State:                      STUCK
Locks OK/REPLICATING/STUCK: 7775/101/617
Grouping:                   DATASET
Expires at:                 None
Locked:                     True
Weight:                     None
Created at:                 2022-11-17 11:39:30
Updated at:                 2022-11-28 12:19:29
Error:                      RequestErrMsg.TRANSFER_FAILED:DESTINATION [17] Destination file exists and overwrite is not enabled
Subscription Id:            None
Source replica expression:  None
Activity:                   User Subscriptions
Comment:                    None
Ignore Quota:               False
Ignore Availability:        False
Purge replicas:             False
Notification:               NO
End of life:                None
Child Rule Id:              None

**Comment by mpatrascoiu** [comment]
Hello,

Can you post the list of the stuck files?
I can check some of them in FTS.

**Comment by maksiks** [comment]
[missing.txt](https://github.com/rucio/rucio/files/10103798/missing.txt)

**Comment by mpatrascoiu** [comment]
I've found the problem!

Taking a look at the first file in that list:
`/store/data/Run2022D/ParkingDoubleMuonLowMass0/RAW/v1/000/357/538/00000/01d5d7e3-4fa1-4dba-bf06-8fc9bf9c9cb0.root`

The FTS job reports `file_on_tape = false / file_on_disk = true` [1]
Using `xrdfs`, the file shows up as being on tape:
```bash
$ xrdfs root://dtn20.nese.rc.fas.harvard.edu:1094/ query prepare `uuidgen` /store/d1ata/Run2022D/ParkingDoubleMuonLowMass0/RAW/v1/000/357/538/00000/01d5d7e3-4fa1-4dba-bf06-8fc9bf9c9cb0.root | jq .responses[] | grep '"on.*"'
  "on_tape": true,
  "online": true,
```

Ideally, FTS should see it as well and mark it as "on tape", upon which Rucio can confirm the replica in the catalogue. So what goes wrong?

When doing the file integrity check, FTS does not use `xrdfs query prepare`, but `xrdfs stat` instead. Here we see a different picture. In the stat request, there is no `BackupExists` flag present to indicate this file is on tape:
```bash
$ gfal-xattr root://dtn20.nese.rc.fas.harvard.edu:1094//store/data/Run2022D/ParkingDoubleMuonLowMass0/RAW/v1/000/357/538/00000/01d5d7e3-4fa1-4dba-bf06-8fc9bf9c9cb0.root user.status
ONLINE

$ xrdfs root://dtn20.nese.rc.fas.harvard.edu:1094/ stat /store/data/Run2022D/ParkingDoubleMuonLowMass0/RAW/v1/000/357/538/00000/01d5d7e3-4fa1-4dba-bf06-8fc9bf9c9cb0.root | grep Flags
Flags:  48 (IsReadable|IsWritable)
```

Let's compare the command output with CTA:
```bash
$ xrdfs root://eosctaatlaspps.cern.ch:1094/ query prepare `uuidgen` /eos/ctaatlaspps/ftstest/IBM1L9/ftstest1/1/test01000000 | jq .responses[] | grep -e '"on.*"'
  "on_tape": true,
  "online": false,

$ gfal-xattr root://eosctaatlaspps.cern.ch:1094//eos/ctaatlaspps/ftstest/IBM1L9/ftstest1/1/test01000000 user.status
NEARLINE

$ xrdfs root://eosctaatlaspps.cern.ch:1094/ stat /eos/ctaatlaspps/ftstest/IBM1L9/ftstest1/1/test01000000 | grep Flags
Flags:  152 (Offline|IsReadable|BackUpExists)
```

In short, the MIT tape endpoint provides the correct response to `xrdfs query prepare`, but does not provide any of the `Offline` and `BackupExists` during the `xrdfs stat` request. For the destination file report (and commonly used in many other frameworks), FTS uses the `gfal-xattr <path> user.status` output (one of `ONLINE`, `NEARLINE` or `ONLINE_AND_NEARLINE`). For the XRoot protocol, the `gfal-xattr <root://path> user.status` interprets the `xrdfs stat` response. [2]

There are two ways to get out of this:
1. The MIT Tape endpoint implements the `Offline` and `BackupExists` stat bits (I suggest this ahs to be done either way, it's more a question of how long it would take here)
2. FTS and Rucio deploy https://github.com/rucio/rucio/issues/5975

Cheers,
Mihai

[1] https://fts3-cms.cern.ch:8449/fts3/ftsmon/#/job/09d66e04-6ee6-11ed-83a5-fa163ebacedb?file=3506270174
[2] https://gitlab.cern.ch/dmc/gfal2/-/blob/v2.21.1/src/plugins/xrootd/gfal_xrootd_plugin_interface.cpp#L289-L315

**Comment by maksiks** [comment]
Ok, here is how the command looks like when I try to execute it:
gfal-xattr root://dtn20.nese.rc.fas.harvard.edu:1094//store/data/Run2022D/ParkingDoubleMuonLowMass0/RAW/v1/000/357/538/00000/01d5d7e3-4fa1-4dba-bf06-8fc9bf9c9cb0.root
xroot.cksum = adler32 39cd11fa
xroot.space = oss.cgroup=all&oss.space=719814918144&oss.free=675833446400&oss.maxf=228051648512&oss.used=43981471744&oss.quota=-1
xroot.xattr = oss.cgroup=public&oss.type=f&oss.used=3704685885&oss.mt=1669169662&oss.ct=1669169662&oss.at=1669169664&oss.u=*&oss.g=*&oss.fs=w&ofs.ap=a
spacetoken = { "totalsize": 1075580070002688, "unusedsize": 882934680649728, "usedsize": 192645389352960, "guaranteedsize": 294311560216576 }


would you please post how the whole thing looks like at CERN?

**Comment by mpatrascoiu** [comment]
I did in my first message. You need to also put `user.status` at the end to get the file locality response. It's one of the `gfal-xattr` reserved keywords

**Comment by maksiks** [comment]
The reason I ask is because I can't repeat any of the commands that you execute for CERN tape:
xrdfs root://eosctaatlaspps.cern.ch:1094/ stat /eos/ctaatlaspps/ftstest/IBM1L9/ftstest1/1/test01000000
[FATAL] Auth failed: No protocols left to try
and you showed xrdfs with grep at the end, it would help if I could see how the whole response looks like.

Anyways, is there a specific version of xrootd that I have to use, or there are some flags that need to be set? Or I have to write custom plugin for xrdfs stat command?

**Comment by mpatrascoiu** [comment]
Ok, I understand. You receive `[FATAL] Auth failed: No protocols left to try` because of permission errors.

Here's the full `xrdfs stat` output against CTA:
```
$ xrdfs root://eosctaatlaspps.cern.ch:1094/ stat /eos/ctaatlaspps/ftstest/IBM1L9/ftstest1/1/test01000000
Path:   /eos/ctaatlaspps/ftstest/IBM1L9/ftstest1/1/test01000000
Id:     6917529027641133823
Size:   1024000
MTime:  2022-10-28 14:34:23
Flags:  152 (Offline|IsReadable|BackUpExists)
```

The `Offline` and `BackUpExists` flags were added in XRootd at the request of the CERN CTA team. At this point, it's best to contact the CTA dev team (cta-dev@cern.ch) or maybe even XRootd developers. I'm not sure you'll get more answers out of this Rucio/FTS thread.

On the FTS side, we will implement a change needed for the second scenario (FTS and Rucio deploy #5975 ), hopefully before the X-Mas break. I need also @rcarpa to confirm the Rucio timelines. However, if you also have the change on your side, then these files will be marked as finished on the next Rucio retry.

I think we've reached the end of this issue. I don't see much more to investigate.

Cheers,
Mihai
